cdap-data-fabric/src/main/java/co/cask/cdap/data/dataset/SystemDatasetInstantiator.java

```diff
@@ -0,0 +1,90 @@
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.data.dataset;
+
+import co.cask.cdap.api.data.DatasetInstantiationException;
+import co.cask.cdap.api.dataset.Dataset;
+import co.cask.cdap.api.dataset.DatasetDefinition;
+import co.cask.cdap.data2.datafabric.dataset.type.DatasetClassLoaderProvider;
+import co.cask.cdap.data2.datafabric.dataset.type.DirectoryClassLoaderProvider;
+import co.cask.cdap.data2.dataset2.DatasetFramework;
+import co.cask.cdap.proto.Id;
+import com.google.common.base.Objects;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.Map;
+import javax.annotation.Nullable;
+
+/**
+ * The data set instantiator creates instances of data sets at runtime. It is mostly a wrapper around
+ * {@link DatasetFramework}, except it is closeable so that any resources created by dataset classloaders
+ * can be properly cleaned up. For example, a {@link DirectoryClassLoaderProvider} will fetch dataset jars
+ * and unpack them in a local directory, which must be cleaned up after the dataset is no longer needed.
+ */
+public class SystemDatasetInstantiator implements Closeable {
+
+  private final DatasetFramework datasetFramework;
+  // provides classloaders to use for different dataset modules
+  private final DatasetClassLoaderProvider classLoaderProvider;
+  private final Iterable<? extends Id> owners;
+  private final ClassLoader parentClassLoader;
+
+  SystemDatasetInstantiator(DatasetFramework datasetFramework,
+                            @Nullable ClassLoader parentClassLoader,
+                            DatasetClassLoaderProvider classLoaderProvider,
+                            @Nullable Iterable<? extends Id> owners) {
+    this.owners = owners;
+    this.classLoaderProvider = classLoaderProvider;
+    this.datasetFramework = datasetFramework;
+    this.parentClassLoader = parentClassLoader == null ?
+      Objects.firstNonNull(Thread.currentThread().getContextClassLoader(), getClass().getClassLoader()) :
+      parentClassLoader;
+  }
+
+  public <T extends Dataset> T getDataset(Id.DatasetInstance datasetId)
+    throws DatasetInstantiationException {
+    return getDataset(datasetId, DatasetDefinition.NO_ARGUMENTS);
+  }
+
+  @SuppressWarnings("unchecked")
+  public <T extends Dataset> T getDataset(Id.DatasetInstance datasetId, Map<String, String> arguments)
+    throws DatasetInstantiationException {
+
+    T dataset;
+    try {
+      if (!datasetFramework.hasInstance(datasetId)) {
+        throw new DatasetInstantiationException("Trying to access dataset that does not exist: " + datasetId);
+      }
+
+      dataset = datasetFramework.getDataset(datasetId, arguments, parentClassLoader, classLoaderProvider, owners);
+      if (dataset == null) {
```

Comment:
this is a move of code that was in DatasetInstantiator. DatasetFramework returns null if the dataset doesn't exist... but that check is made above.  Should be safe to remove the hasInstance check and replace this exception with that one.